br/pkg/task/restore.go

```diff
@@ -1942,3 +1984,162 @@ func checkIsInActions(action model.ActionType, actions map[model.ActionType]stru
+		return errors.Annotate(berrors.ErrRestoreModeMismatch, "cannot do transactional restore from raw kv data")
+	}
+
+	// verify dbs and tables are in backup
+	if err := VerifyDBAndTableInBackup(client.GetDatabases(), cfg.RestoreConfig); err != nil {
+		return errors.Trace(err)
+	}
+
+	if isFullRestore(cmdName) {
+		if client.NeedCheckFreshCluster(cfg.ExplicitFilter, checkpointEnabledAndExists) {
+			if err := client.EnsureNoUserTables(); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	return nil
+}
+
+// checkOptionalClusterRequirements checks disk space and table existence. It's optional since it can be turned off by
+// config cfg.CheckRequirements
+func checkOptionalClusterRequirements(
+	ctx context.Context,
+	client *snapclient.SnapClient,
+	cfg *SnapshotRestoreConfig,
+	checkpointEnabledAndExists bool,
+	mgr *conn.Mgr,
+	files []*backuppb.File,
+	tables []*metautil.Table,
+	isPitr bool) error {
+	if cfg.CheckRequirements && !checkpointEnabledAndExists {
+		if err := checkDiskSpace(ctx, mgr, files, tables); err != nil {
+			return errors.Trace(err)
+		}
+		if !client.IsIncremental() {
+			if err := checkTableExistence(ctx, mgr, tables); err != nil {
+				return errors.Trace(err)
+			}
+		}
+		if isPitr {
+			if err := checkTableExistence(ctx, mgr, buildLogBackupMetaTables(cfg.PiTRTableTracker.DBNameToTableNames)); err != nil {
+				return errors.Trace(err)
+			}
+		}
+	}
+	return nil
+}
+
+func buildLogBackupMetaTables(dbNameToTableNames map[string]map[string]struct{}) []*metautil.Table {
+	tables := make([]*metautil.Table, 0)
+
+	for dbName, tableNames := range dbNameToTableNames {
+		for tableName := range tableNames {
+			table := &metautil.Table{
+				DB: &model.DBInfo{
+					Name: ast.NewCIStr(dbName),
+				},
+				Info: &model.TableInfo{
+					Name: ast.NewCIStr(tableName),
+				},
+			}
+			tables = append(tables, table)
+		}
+	}
+	return tables
+}
+
+func createDBsAndTables(
+	ctx context.Context,
+	client *snapclient.SnapClient,
+	cfg *SnapshotRestoreConfig,
+	mgr *conn.Mgr,
+	dbs []*metautil.Database,
+	tables []*metautil.Table,
+	isPiTR bool) ([]*snapclient.CreatedTable, error) {
+	var newTS uint64
+	if client.IsIncremental() {
+		if !cfg.AllowPITRFromIncremental {
+			// we need to get the new ts after execDDL
+			// or backfilled data in upstream may not be covered by
+			// the new ts.
+			// see https://github.com/pingcap/tidb/issues/54426
+			var err error
+			newTS, err = restore.GetTSWithRetry(ctx, mgr.GetPDClient())
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+		}
+	}
+
+	// create databases first, it will skip if already exists
+	if err := client.CreateDatabases(ctx, dbs); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	createdTables, err := client.CreateTables(ctx, tables, newTS)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// update table mapping manager with new table ids if pitr
+	if isPiTR {
+		cfg.tableMappingManager.UpdateDownstreamIds(tables, client.GetDomain())
```

Comment:
`UpdateDownstreamIds` doesn't need to get tableInfo from domain because `createdTables` already keeps them. See https://github.com/pingcap/tidb/blob/f8f667ed816f1bb79d30c98d49491a518fcbbc48/br/pkg/restore/snap_client/client.go#L917